src/osp/Active/SysPrefabInit.cpp

```diff
@@ -41,35 +42,43 @@ using osp::phys::EShape;
 namespace osp::active
 {
 
-void SysPrefabInit::init_hierarchy(
-        ACtxPrefabInit const&               rPrefabInit,
+void SysPrefabInit::add_to_subtree(
+        TmpPrefabInitBasic const&           basic,
+        ArrayView<ActiveEnt const>          ents,
         Resources const&                    rResources,
-        acomp_storage_t<ACompHierarchy>&    rHier) noexcept
+        SubtreeBuilder&                     bldPrefab) noexcept
 {
-    auto itPfEnts = std::begin(rPrefabInit.m_ents);
+    auto const &rImportData = rResources.data_get<osp::ImporterData const>(
+            gc_importer, basic.m_importerRes);
+    auto const &rPrefabData = rResources.data_get<osp::Prefabs const>(
+            gc_importer, basic.m_importerRes);
 
-    for (TmpPrefabInitBasic const& rPfBasic : rPrefabInit.m_basic)
-    {
-        auto const &rPrefabData = rResources.data_get<osp::Prefabs const>(
-                gc_importer, rPfBasic.m_importerRes);
+    auto const objs     = rPrefabData.m_prefabs[basic.m_prefabId];
+    auto const parents  = rPrefabData.m_prefabParents[basic.m_prefabId];
 
-        std::size_t const objCount = rPrefabData.m_prefabs[rPfBasic.m_prefabId].size();
-        auto const parents = rPrefabData.m_prefabParents[rPfBasic.m_prefabId];
+    auto itObj      = std::begin(objs);
+    auto itEnt      = std::begin(ents);
 
-        for (std::size_t i = 0; i < objCount; ++i)
-        {
-            ObjId const prefabParent = parents[i];
-            ActiveEnt const parent = (prefabParent == -1)
-                    ? rPfBasic.m_parent
-                    : (*itPfEnts)[prefabParent];
+    auto const add_child_recurse
+            = [&rImportData, &itObj, &itEnt, &parents] (auto&& self, SubtreeBuilder& bldParent) -> void
```

Comment:
conceptually, since this lambda modifies things it points to, you would probably want to mark it as mutable.